nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJsonTest.java

```diff
@@ -0,0 +1,409 @@
+
+        clientService.setEvalParametersConsumer((Map<String, String> params) -> {
+            assertEquals(2, params.size());
+            assertEquals("true", params.get("refresh"));
+            assertEquals("auto", params.get("slices"));
+        });
+
+        clientService.setEvalConsumer((List<IndexOperationRequest> items) -> {
+            long nullIdCount = items.stream().filter(item -> item.getId() == null).count();
+            long headerFieldsCount = items.stream().filter(item -> !item.getHeaderFields().isEmpty()).count();
+            assertEquals(1L, nullIdCount);
+            assertEquals(1L, headerFieldsCount);
+
+            Map<String, String> headerFields = items.get(0).getHeaderFields();
+            assertEquals(2, headerFields.size());
+            assertEquals("1", headerFields.get("routing"));
+            assertEquals("external", headerFields.get("version"));
+        });
+
+        Map<String, String> attributes = new LinkedHashMap<>();
+        attributes.put("slices", "auto");
+        attributes.put("version", "external");
+        attributes.put("blank", " ");
+        attributes.put("doc_id", "");
+        basicTest(0, 0, 1, attributes);
+    }
+
+    @Test
+    public void simpleTestWithScriptAndDynamicTemplates() throws Exception {
+        runner.setProperty(PutElasticsearchJson.SCRIPT, Files.readString(Paths.get(TEST_DIR,"script.json")));
+        runner.setProperty(PutElasticsearchJson.DYNAMIC_TEMPLATES, Files.readString(Paths.get(TEST_COMMON_DIR,"dynamicTemplates.json")));
+        final Map<String, Object> expectedScript = new LinkedHashMap<>();
+        expectedScript.put("_source", "some script");
+        expectedScript.put("language", "painless");
+        final Map<String, Object> expectedDynamicTemplate = new LinkedHashMap<>();
+        expectedDynamicTemplate.put("my_field", "keyword");
+        final Map<String, Object> yourField = new LinkedHashMap<>();
+        yourField.put("type", "text");
+        yourField.put("keyword", Collections.singletonMap("type", "text"));
+        expectedDynamicTemplate.put("your_field", yourField);
+        Consumer<List<IndexOperationRequest>> consumer = (List<IndexOperationRequest> items) -> {
+            long scriptCount = items.stream().filter(item -> item.getScript().equals(expectedScript)).count();
+            long falseScriptedUpsertCount = items.stream().filter(item -> !item.isScriptedUpsert()).count();
+            long dynamicTemplatesCount = items.stream().filter(item -> item.getDynamicTemplates().equals(expectedDynamicTemplate)).count();
+            assertEquals(1L, scriptCount);
+            assertEquals(1L, falseScriptedUpsertCount);
+            assertEquals(1L, dynamicTemplatesCount);
+        };
+        basicTest(0, 0, 1, consumer, null);
+
+        runner.clearTransferState();
+        runner.clearProvenanceEvents();
+
+        runner.setProperty(PutElasticsearchJson.INDEX_OP, IndexOperationRequest.Operation.Upsert.getValue().toLowerCase());
```

Comment:
`simpleTestWithScriptedUpsert`

Reply:
Should line 260-266 be its own unit test? If so what should it be named? Perhaps simpleTestWithNonJsonScriptedUpsert.

Reply:
Sounds like a sensible refactor, maybe `testInvalidScript` or `testNonJsonScript`?